### ISSUE TITLE
Shoryuken does not finish all jobs before shutdown

### ISSUE THREAD

**Issue by erozycki** [phstc/shoryuken#539]
When I do the following:

1. Saturate a queue with extremely short jobs,
2. Start shoryuken,
3. Stop shoryuken with USR1 or TERM,

I end up with a non-zero number of messages in flight, as indicated in the SQS console, after Shoryuken shuts down.

One guess is that Shoryuken waits for workers to finish, but it does not wait for an in-flight SQS.receive_message_batch() to return before shutting down. When I add a short, random sleep(0.2 + rand) to my jobs I no longer experience this issue, possibly because by lengthening the execution time of these jobs I ensure that any in-flight receive_message_batch() returns before all currently executing jobs finish.

I'm using ruby 2.5.1, shoryuken 4.0.2.

:pushpin: <s>#560</s>

**Comment by phstc** [comment]
@erozycki that makes sense.

I think the fix would be to when stopping, we could iterate over all managers and turn `running` into false.

Ref:

* https://github.com/phstc/shoryuken/blob/352b20642b4b8a123b821228231ed5a5c618a9cc/lib/shoryuken/launcher.rb#L16-L33
* https://github.com/phstc/shoryuken/blob/352b20642b4b8a123b821228231ed5a5c618a9cc/lib/shoryuken/manager.rb#L15
* https://github.com/phstc/shoryuken/blob/352b20642b4b8a123b821228231ed5a5c618a9cc/lib/shoryuken/manager.rb#L109

Could you try that? ❤️

**Comment by erozycki** [comment]
Would that fix it? What if a manager sends a request to sqs immediately before the stop, but the response doesn't come until some time after?

I also tried that fix by defining

```
def stop
  @running.make_false
end
```
on `Manager` and adding `managers.each(&:stop)` as the first line of `stop!`, but I still encounter the issue. (I may have misunderstood the instructions.)

**Comment by phstc** [comment]
_![image](https://avatars1.githubusercontent.com/u/105652?s=22&v=4) is thinking out loud..._ 


@erozycki if you send `TERM`, there will be always a chance of leaving in-flight messages, [since the process will eventually hard shutdown giving a timeout](https://github.com/phstc/shoryuken/blob/352b20642b4b8a123b821228231ed5a5c618a9cc/lib/shoryuken/launcher.rb#L21).

The idea of setting running to false, was to make [this](https://github.com/phstc/shoryuken/blob/352b20642b4b8a123b821228231ed5a5c618a9cc/lib/shoryuken/manager.rb#L35) to return false and prevent a future [`dispatch`](https://github.com/phstc/shoryuken/blob/352b20642b4b8a123b821228231ed5a5c618a9cc/lib/shoryuken/manager.rb#L35).

But checking the concurrent-ruby docs, it seems `@executor.running?` should be false anyway: 

> false when shutting down or shutdown
> https://ruby-concurrency.github.io/concurrent-ruby/master/Concurrent/ThreadPoolExecutor.html#running%3F-instance_method


But reviewing the code, I think [this](https://github.com/phstc/shoryuken/blob/352b20642b4b8a123b821228231ed5a5c618a9cc/lib/shoryuken/manager.rb#L65) might be the problem, `assign` could always try to `assign`.

Can you try to revert the changes and only remove [this](https://github.com/phstc/shoryuken/blob/352b20642b4b8a123b821228231ed5a5c618a9cc/lib/shoryuken/manager.rb#L65) this line instead?

Sorry, I can't test it right now. If you can test it, that would be great!

**Comment by phstc** [comment]
@erozycki sorry for the confusion, after a while I couldn't follow my own instructions above.

I created PR #560, could you test it?

> 1. Saturate a queue with extremely short jobs,
> 2. Start shoryuken,
> 3. Stop shoryuken with USR1 or TERM,

**Comment by erozycki** [comment]
Great, thanks, will check this out soon!

**Comment by erozycki** [comment]
I just tested, and I don't think this (entirely) resolves the issue. I did steps 1, 2, and 3 (used USR1), and afterward I would be left with > 0 messages in flight. I tried five times, and after each try I was left with more messages in flight than before. Each time I waited a while and refreshed the SQS dashboard several times.

At the same time, I am not very concerned about this issue. After all, in general there is no guarantee that a message dequeued from SQS will be processed successfully. So even if this specific issue were mitigated or resolved, we still would to have to retry jobs occasionally after waiting for the visibility timeout to elapse. Nowadays I use a shorter visibility timeout than I used to, which helps mitigate the impact of this issue in my use case.

Since this issue isn't causing me trouble, I have no objection to closing it.

Also want to say thanks for the time you've put into Shoryuken, and into investigating this issue!
